### PR TITLE
Refactor: Stop Splitting Header Values

### PR DESCRIPTION
**Pull request by Johannes-Schneider**
## Context
 

SAP/cloud-sdk-java-backlog#156.
SAP/cloud-sdk-java-backlog/discussions/343

This PR changes our `RequestHeaderContainer` and the `ODataRequestResult` to no longer split header values.

### Feature scope:

 
- [x] Adjust `RequestHeaderContainer`
  - [x] JavaDoc
  - [x] Implementation (`DefaultRequestHeaderContainer`)
- [x] Adjust `ODataRequestResult` 

## Definition of Done

- [x] Functionality scope stated & covered
- [x] Tests cover the scope above
- [x] ~Error handling created / updated & covered by the tests above~
- [x] Documentation updated → https://github.com/SAP/cloud-sdk/pull/1590
- [x] Release notes updated

### CODE REVIEW

File: cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java [hunk 1]
```diff
@@ -214,7 +204,7 @@ public RequestHeaderContainer.Builder withHeader(
         @Override
         public RequestHeaderContainer.Builder withoutHeader( @Nonnull final String name )
         {
-            headers.removeAll(normalize(name));
+            headers.remove(normalize(name));
```
Comment (newtork) [inline cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java:207]: (Comment/Question)

Why did we decide to "normalize" header name?
Btw, I see no case-insensitivity. Might be a problem in/of the future.

Comment (newtork) [reply cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java:207]: I got confused by your introduced `normalize( Iterable<String> )`

File: cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java [hunk 2]
```diff
@@ -81,12 +81,7 @@ public static RequestHeaderContainer fromMultiValueMap(
                 continue;
             }
 
-            // Note: Currently customers are unable to influence the way header values are split.
-            // This might become an issue in the future, for example, if the system doesn't obey to the HTTP specification (e.g. comma "," should not be used to split header values).
-            // To enable such edge-cases, we might consider adding a factory for the "RequestHeaderCollection" where customers would be able to apply their own splitting logic.
-            // Such a factory could also be used to inject/remove headers from the initial servlet request.
-
-            normalizedResult.putAll(normalize(headerName), splitHeaderValues(headerName, headerValues));
+            normalizedResult.putAll(normalize(headerName), normalize(headerValues));
```
Comment (newtork) [inline cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java:84]: (Minor)

Not sure whether this makes it more readable than having a dedicated helper method 🤔 

```suggestion
            normalizedResult.putAll(normalize(headerName), Iterables.filter(headerValues, Objects::nonNull));
```

File: cloudplatform/cloudplatform-core/src/test/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainerTest.java [hunk 3]
```diff
@@ -89,65 +89,54 @@ public void creationRemovesNullAndEmptyValues()
...
-        input
-            .put("sEt-CoOkIe", Arrays.asList("cookie5=value5", "cookie6=value6;cookie7=value, which contains a comma"));
+        input.put("Header1", Collections.singletonList("  Value1-1   "));
+        input.put("Header2", Collections.singletonList("  Value2-1  ,  Value2-2   "));
+        input.put("Set-Cookie", Collections.singletonList("  cookie1  = cookie-value1  "));
+        input.put("set-cookie", Collections.singletonList("  cookie2 = cookie-value2  ; cookie3  "));
 
         final RequestHeaderContainer sut = DefaultRequestHeaderContainer.fromMultiValueMap(input);
 
+        assertThat(sut.getHeaderValues("Header1")).containsExactlyInAnyOrder("  Value1-1   ");
```
Comment (newtork) [inline cloudplatform/cloudplatform-core/src/test/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainerTest.java:125]: (Comment)

I'm wondering whether it'd be worth it investigating how Spring/Apache/JDK11 `HttpClient` is handling white spaces in response header values-

File: datamodel/odata-client/src/test/java/com/sap/cloud/sdk/datamodel/odata/client/request/ODataRequestResultTest.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        when(sut.getAllHeaderValues()).thenCallRealMethod();
+
+        final Map<String, Iterable<String>> actual = sut.getAllHeaderValues();
+
+        assertThat(actual).containsExactly(entry("Header", Arrays.asList("Value1-1", "Value1-2")));
+    }
+
+    @SafeVarargs
+    @Nonnull
+    private static HttpResponse mockResponseWithHeaders(
```
Comment (newtork) [inline datamodel/odata-client/src/test/java/com/sap/cloud/sdk/datamodel/odata/client/request/ODataRequestResultTest.java:86]: (Minor)

Feels like this method can be nicely reduced by just using implementation `BasicHttpResponse` with respective string-based header-methods.

**Review by newtork — APPROVED** [review]
Only minor comments, otherwise LGTM

**Review by cschubertcs — COMMENTED** [review]
lgtm!
When we offer the same public API we should also have a look the concrete implementation. Best to use the same kind of Map (ImmutableListMultimap with "normalized" keys vs. TreeMap with `String::compareToIgnoreCase`). But for the moment that should result in the same behavior.

File: cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java [hunk 5]
```diff
@@ -32,15 +31,8 @@
 @Beta
 @EqualsAndHashCode
 @RequiredArgsConstructor( access = AccessLevel.PRIVATE )
-public class DefaultRequestHeaderContainer implements RequestHeaderContainer
+public final class DefaultRequestHeaderContainer implements RequestHeaderContainer
```
Comment (newtork) [inline cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java:34]: (Question)

Why final?

**Review by newtork — APPROVED** [review]
LGTM

Very minor, map implementation choice as Chris mentioned. Minor code quality.

File: cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java (hunk 5 above)
Comment (Johannes-Schneider) [reply cloudplatform/cloudplatform-core/src/main/java/com/sap/cloud/sdk/cloudplatform/requestheader/DefaultRequestHeaderContainer.java:34]: As discussed, the constructor is `private` so the class cannot be inherited (in productive code).

For mocking (i.e. during tests), I'd expect customers to either mock the interface (`RequestHeaderContainer`) or to use an actual (i.e. non-mocked) instance of the implementation (`DefaultRequestHeaderContainer`).
Therefore, I'd argue this change to be fine.

**Pull request merged**